### PR TITLE
Hotfix/re indexing

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

New endpoint to enable the api the ability (if authenticated) to kick off a search build of a particular hierarchy (dimension) of an instance.

### How to review

Already been reviewed and merged into cmd-develop, and tested in environment.

### Who can review

Anyone

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,6 +35,9 @@ one of:
 | GRACEFUL_SHUTDOWN_TIMEOUT  | 5s                                   | The graceful shutdown timeout
 | HEALTHCHECK_INTERVAL       | 1m                                   | The time between calling the healthcheck endpoint for check subsystems
 | HEALTHCHECK_TIMEOUT        | 2s                                   | The timeout that the healthcheck allows for checked subsystems
+| HIERARCHY_BUILT_TOPIC      | hierarchy-built                      | The kafka topic to write messages to
+| KAFKA_ADDR                 | "localhost:9092"                     | The list of kafka hosts
```
Comment (gedge) [inline README.md]: why quote this and not others?

File: api/api.go [hunk 2]
```diff
@@ -5,6 +5,7 @@ import (
 
 	"github.com/ONSdigital/dp-dataset-api/store"
 	"github.com/ONSdigital/dp-search-api/auth"
+	"github.com/ONSdigital/dp-search-api/searchOutputQueue"
```
Comment (gedge) [inline api/api.go]: what mixed-case hell is this? 😄

File: api/search_test.go [hunk 3]
```diff
@@ -91,82 +91,136 @@ func TestGetSearchFailureScenarios(t *testing.T) {
...
 
 	Convey("Given the offset parameter exceeds the default maximum results return status 400 (bad request)", t, func() {
 		r := httptest.NewRequest("GET", "http://localhost:23100/search/datasets/123/editions/2017/versions/1/dimensions/aggregate?q=term&offset=50", nil)
 		w := httptest.NewRecorder()
 
-		api := routes(host, secretKey, datasetAPISecretKey, mux.NewRouter(), &mocks.DatasetAPI{}, &mocks.Elasticsearch{}, defaultMaxResults)
+		api := routes(host, secretKey, datasetAPISecretKey, mux.NewRouter(), &mocks.BuildSearch{}, &mocks.DatasetAPI{}, &mocks.Elasticsearch{}, defaultMaxResults)
 		api.router.ServeHTTP(w, r)
 		So(w.Code, ShouldEqual, http.StatusBadRequest)
+		So(w.Body.String(), ShouldEqual, "the maximum offset has been reached, the offset cannot be more than 1000\n")
```
Comment (gedge) [inline api/search_test.go]: but offset is not more than 1000 😕

File: api/search_test.go [hunk 4]
```diff
@@ -91,82 +91,136 @@ func TestGetSearchFailureScenarios(t *testing.T) {
...
+
+func TestFailToCreateSearchIndex(t *testing.T) {
+	Convey("Given a request to create search index but no auth header is set return a status 401 (unauthorised)", t, func() {
+		r := httptest.NewRequest("PUT", "http://localhost:23100/search/instances/123/dimensions/aggregate", nil)
+		w := httptest.NewRecorder()
+
+		api := routes(host, secretKey, datasetAPISecretKey, mux.NewRouter(), &mocks.BuildSearch{}, &mocks.DatasetAPI{}, &mocks.Elasticsearch{}, defaultMaxResults)
+		api.router.ServeHTTP(w, r)
+		So(w.Code, ShouldEqual, http.StatusUnauthorized)
+		So(w.Body.String(), ShouldEqual, "No authentication header provided\n")
```
Comment (gedge) [inline api/search_test.go]: personally, I'd remove the word 'header' here, because a12n may use other ways in the future

File: config/config.go [hunk 5]
```diff
@@ -10,12 +10,15 @@ import (
...
 	BindAddr                  string        `envconfig:"BIND_ADDR"                  json:"-"`
+	Brokers                   []string      `envconfig:"KAFKA_ADDR"                 json:"-"`
 	DatasetAPIURL             string        `envconfig:"DATASET_API_URL"`
 	DatasetAPISecretKey       string        `envconfig:"DATASET_API_SECRET_KEY"     json:"-"`
 	ElasticSearchAPIURL       string        `envconfig:"ELASTIC_SEARCH_URL"         json:"-"`
 	GracefulShutdownTimeout   time.Duration `envconfig:"GRACEFUL_SHUTDOWN_TIMEOUT"`
 	HealthCheckInterval       time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
 	HealthCheckTimeout        time.Duration `envconfig:"HEALTHCHECK_TIMEOUT"`
+	HierarchyBuiltTopic       string        `envconfig:"HIERARCHY_BUILT_TOPIC"`
+	KafkaMaxBytes             string        `envconfig:"KAFKA_MAX_BYTES"`
```
Comment (gedge) [inline config/config.go]: why not an `int` or `int64`?

File: main.go [hunk 6]
```diff
@@ -3,13 +3,17 @@ package main
...
 
 	"github.com/ONSdigital/dp-search-api/config"
 	"github.com/ONSdigital/dp-search-api/dataset"
 	"github.com/ONSdigital/dp-search-api/elasticsearch"
+	"github.com/ONSdigital/dp-search-api/searchOutputQueue"
 	"github.com/ONSdigital/dp-search-api/service"
+	"github.com/ONSdigital/go-ns/kafka"
 	"github.com/ONSdigital/go-ns/log"
 	"github.com/ONSdigital/go-ns/rchttp"
+	"github.com/pkg/errors"
```
Comment (gedge) [inline main.go]: why?

File: README.md (hunk 1 above)
Comment (nshumoogum) [reply README.md]: copy pastor job

File: config/config.go (hunk 5 above)
Comment (nshumoogum) [reply config/config.go]: To be consistent with other repositories, (filter API used string)

File: api/search_test.go (hunk 4 above)
Comment (nshumoogum) [reply api/search_test.go]: I wont change this a this will return not found in the future (to hide auth endpoints)

File: api/search_test.go (hunk 3 above)
Comment (nshumoogum) [reply api/search_test.go]: Good spot

File: config/config.go (hunk 5 above)
Comment (nshumoogum) [reply config/config.go]: I will change it and update dp-ci with the new configuration value

**Pull request merged**